apollo-router/src/plugins/connectors/handle_responses.rs

```diff
@@ -597,17 +601,59 @@ async fn deserialize_response<T: HttpBody>(
         }
     }
 
-    match serde_json::from_slice::<Value>(body) {
-        Ok(json_data) => Ok(json_data),
-        Err(_) => {
+    let content_type = parts
+        .headers
+        .get(CONTENT_TYPE)
+        .and_then(|h| h.to_str().ok())
+        .and_then(|s| s.parse::<Mime>().ok());
+
+    if content_type.is_none()
+        || content_type
+            .as_ref()
+            .is_some_and(|ct| ct.subtype() == mime::JSON || ct.suffix() == Some(mime::JSON))
+    {
+        // Treat any JSON-y like content types as JSON
+        // Also, because the HTTP spec says we should effectively "guess" the content type if there is no content type (None), we're going to guess it is JSON if the server has not specified one
+        match serde_json::from_slice::<Value>(body) {
+            Ok(json_data) => Ok(json_data),
+            Err(_) => {
+                if let Some(debug_context) = debug_context {
+                    debug_context
+                        .lock()
+                        .push_invalid_response(debug_request.clone(), parts, body);
+                }
+                Err(make_err(path))
+            }
+        }
+    } else if content_type
+        .as_ref()
+        .is_some_and(|ct| ct.type_() == mime::TEXT && ct.subtype() == mime::PLAIN)
+    {
+        // Plain text we can't parse as JSON so we'll instead return it as a JSON string
+        // Before we can do that, we need to figure out the charset and attempt to decode the string
+        let charset = content_type
+            .as_ref()
+            .and_then(|ct| {
+                ct.get_param("charset")
+                    .map(|c| c.as_str().to_ascii_lowercase())
+            })
+            .unwrap_or("utf-8".to_string());
+        let encoding = Encoding::for_label(charset.as_bytes()).unwrap_or(UTF_8);
```

Comment:
```suggestion
        let encoding = content_type
            .as_ref()
            .and_then(|ct| Encoding::for_label(ct.get_param("charset")?.as_str().as_bytes()))
            .unwrap_or(UTF_8);
```

We can avoid double-declaring the default _and_ `Encoding::for_label` does the lowercasing for us, so we can avoid the allocation.